TRUNK-5338 Add changeset to create or modify conditions table

paradisekelechi opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111 Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
Added liquibase changeset to add a new conditions table or modify an existing one. 

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-5338

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
@paradisekelechi did you actually test this out by running an instance of openmrs, go through the database setup wizard, complete it, check the database to confirm all is changed as expected?

paradisekelechi commented:
@dkayiwa 
I actually tested it. I ran it an the changes were effected accordingly on the already existing `conditions` table. After that, I dropped the table and ran and checked the other changeset. 

However, I do not know about the database setup wizard.  Could you share resources with respect to that?

dkayiwa commented:
In the liquibasechangelog table, delete the row with this changeset id. Then restart openmrs and tell me what you get in the browser.

paradisekelechi commented:
About that, I did that a lot of times until I no longer have errors on the console. 
I'd run that again and update you accordingly.

dkayiwa commented:
How are you dealing with the data migration? Didn't some columns store data in a different format than  that of the old table?

paradisekelechi commented:
The only tables that changed are `previous_condition_id`, `status` and `concept_id` and they only changed in name and not datatype.

paradisekelechi commented:
What I am only concerned about is adding foreign key constraints after converting an old conditions table to the new one. . .

dkayiwa commented:
1) Can you revert to the old table.
2) Create around 5 conditions from the user interface when running on openmrs-core versions 2.1.3 or lower.
3) Now switch to openmrs-core 2.2.0-SNAPSHOT to have your changesets run
4) Check to see if the data is migrated and expected and also displayed in the UI using the latest snapshot versions of the ermapi and coreapps modules.

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,29 @@
+		<comment>
+			Modify conditions table to match condition object in the core platform if the conditions table already exists in the database
+		</comment>
+		<renameColumn columnDataType="int" newColumnName="previous_version" oldColumnName="previous_condition_id" tableName="conditions"/>
```
I would prefer to have these as separate changesets with their respective comments being explicit about what they do e.g **Rename tableA.x column to y**

paradisekelechi replied:
Okay. Done

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,29 @@
+		<renameColumn columnDataType="varchar(255)" newColumnName="clinical_status" oldColumnName="status" tableName="conditions"/>
+		<renameColumn columnDataType="int" newColumnName="condition_coded" oldColumnName="concept_id" tableName="conditions"/>
+
+		<addColumn tableName="conditions">
```
This should also be a separate changeset

paradisekelechi replied:
Okay. Noted

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,29 @@
+			<column name="condition_coded_name" type="int"/>
+		</addColumn>
+	</changeSet>
 	<changeSet id="20180214-1000" author="Kelechi+iheanyi">
```
It's good to change the id when you change the changeset contents so that this doesn't break other developer's working copies or our CI servers

paradisekelechi replied:
Yes. Done

wluyima replied:
I don't see the change

paradisekelechi commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,29 @@
 		</preConditions>
 		<comment>
-			Add condition table
+			Add conditions table if the table does not exist in the database
```
Okay. Noted

dkayiwa commented:
@paradisekelechi have you tried to run a fresh setup of openmrs with these changes?

dkayiwa commented:
@paradisekelechi if you do, you should get this error: https://pastebin.com/QxxXu5h6

paradisekelechi commented:
@dkayiwa
I have run a fresh setup of the core platform following the steps you highlighted. I did not get that error. The changesets ran successfully without errors

dkayiwa commented:
Do you have the reference application modules loaded?

paradisekelechi commented:
Yes I do have the reference application modules loaded

dkayiwa commented:
That does not look like a fresh setup. For the emrapi module would attempt to create a conditions table which should fail because the core platform already created it.

paradisekelechi commented:
@dkayiwa 
I have done the above. This is the output I get in the console for these changesets: https://pastebin.com/Fr0w4nt9 .
On checking the db created, the `conditions` table is there with the changes effected accordingly.

paradisekelechi commented:
1. I dropped the previous databases I had.
2. I deleted the openmrs-runtime.properties file
3. I started the server and followed through on the steps which created a new db and updated it by running the changesets

dkayiwa commented:
As you did that, which version of openmrs-core were you running?

paradisekelechi commented:
2.2.0-SNAPSHOT

dkayiwa commented:
And when you run the setup, did you have all the reference application modules loaded from the very start?

paradisekelechi commented:
No I did not have them

dkayiwa commented:
@paradisekelechi can you do that?

dkayiwa commented:
Now do a fresh setup with those modules installed

samuelmale commented:
Are you still working on this @paradisekelechi ?

samuelmale commented:
Are you still working on this @paradisekelechi ?

paradisekelechi commented:
I have worked on this. I am waiting for verifications and updates from @dkayiwa

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,23 @@
 		</preConditions>
 		<comment>
-			Add condition table
+			Rename conditions table to emrapi_conditions
```
This comments seems to be saying the opposite

wluyima commented on api/src/main/resources/liquibase-update-to-2.2.xml:
```diff
@@ -117,14 +117,23 @@
 		</preConditions>
 		<comment>
-			Add condition table
+			Rename conditions table to emrapi_conditions
```
Why are you renaming the table, I thought you should be updating it to match the new design

paradisekelechi replied:
@wluyima 
We are renaming the table because of some edge cases we saw in the implementation. 
Assume we are running a fresh instance of the core platform and the emrapi module.
If we update the `conditions` table here, the emrapi module tries to create a new `conditions` table which throws errors.

However, the following implementation handles that edge case successfully. On the core platform, the conditions table is renamed to emrapi_conditions. Then a conditions table is created and data migrated to it. On the emrapi module, the emrapi_conditions table is used.

wluyima replied:
You should not be renaming the table, in any case it's no longer going to live in the emrapi module so there if no need to add that prefix

mozzy11 replied:
@wluyima  , wat do u mean if the table exists , just do the migration ..? i think that will be just do the modification.  because migration is posiible only after first renaming it.

paradisekelechi commented:
I have thoroughly tested this @dkayiwa

samuelmale commented:
Hi @paradisekelechi 
Are still working on this? Is so, did you address all @wluyima's  requested changes? 
Please look into this and we close this :)
